signac/core/file_buffered_synced_collection.py

```diff
@@ -0,0 +1,256 @@
+        """
+        return json.dumps(self.to_base()).encode()
+
+    @staticmethod
+    def _decode(blob):
+        """Decode serialized data.
+
+        Mirroring _encode, this method assumes JSON serialization.
+
+        Parameters
+        ----------
+        blob : bytes
+            Byte literal to be decoded.
+        """
+        return json.loads(blob.decode())
+
+    def _sync_buffer(self):
+        """Store data in buffer.
+
+        We can reasonably provide a default implementation for all file-based
+        backends that simply entails storing data to an in-memory cache (which
+        could also be a Redis instance, etc).
+        """
+        if self._filename in self._cache:
+            blob = self._encode()
+            cached_data = self._cache[self._filename]
+            buffer_size_change = sys.getsizeof(blob) - sys.getsizeof(
+                cached_data["contents"]
+            )
+            FileBufferedCollection._CURRENT_BUFFER_SIZE += buffer_size_change
+            cached_data["contents"] = blob
+        else:
+            self._initialize_data_in_cache()
+
+        if (
+            FileBufferedCollection._CURRENT_BUFFER_SIZE
+            > FileBufferedCollection._BUFFER_CAPACITY
+        ):
+            FileBufferedCollection._flush_buffer(force=True)
+        # If multiple collections point to the same data, just checking that
+        # the file contents are cached is not a sufficient check.
+        if id(self) not in self._cached_collections:
+            self._cached_collections[id(self)] = self
+
+    def _load_buffer(self):
+        """Read data from buffer.
+
+        We can reasonably provide a default implementation for all file-based
+        backends that simply entails reading data from an in-memory cache
+        (which could also be a Redis instance, etc).
+
+        Returns
+        -------
+        Collection
+            A collection of the same base type as the SyncedCollection this
+            method is called for, corresponding to data loaded from the
+            underlying file.
+        """
+        if self._filename in self._cache:
+            # If multiple collections point to the same data, just checking
+            # that the file contents are cached is not a sufficient check.
+            if id(self) not in self._cached_collections:
+                self._cached_collections[id(self)] = self
+        else:
+            self._initialize_data_in_cache()
+
+        # Load from buffer
+        blob = self._cache[self._filename]["contents"]
+
+        if (
+            FileBufferedCollection._CURRENT_BUFFER_SIZE
+            > FileBufferedCollection._BUFFER_CAPACITY
+        ):
+            FileBufferedCollection._flush_buffer(force=True)
+        return self._decode(blob)
+
+    def _initialize_data_in_cache(self):
+        """Create the initial entry for the data in the cache."""
+        data = self.to_base()
+        blob = json.dumps(data).encode()
```

Comment:
Does this need to be a call to `_encode`, or is that a separate thing that happens to match the implementation here?